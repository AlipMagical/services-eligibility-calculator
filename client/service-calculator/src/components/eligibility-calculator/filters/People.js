/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Col, Row } from 'react-bootstrap';
import React, { useState } from 'react';

import { Chip } from '../Chip';
import FilterModal from './FilterModal';
import PropTypes from 'prop-types';
import ReactGA from 'react-ga';
import { capitalize } from '../../../common/utils';
import { getTotalPeople } from '../../../filters/filterAge';
import { useTranslation } from 'react-i18next';

function PersonInput({ label, description, onChange, value }) {
  const { t } = useTranslation();
  return (
    <Row className="my-3">
      <Col xs={6}>
        <span className="input-label">{label}</span>
        <span className="people-input-description">{description}</span>
      </Col>
      <Col xs={6} className="people-input">
        <button
          className="btn btn-people-input"
          onClick={() => onChange(value - 1)}
          disabled={value < 1}
          aria-label={`${t('filterPeople.buttonDecreaseLabel')} ${label}`}
        >
          <i className="material-icons align-middle">remove</i>
        </button>
        <span className="people-input-value align-middle">{value}</span>
        <button
          className="btn btn-people-input"
          onClick={() => onChange(value + 1)}
          aria-label={`${t('filterPeople.buttonIncreaseLabel')} ${label}`}
        >
          <i className="material-icons align-middle">add</i>
        </button>
      </Col>
    </Row>
  );
}

function sendAnalytics(people) {
  let totalPeople = 0;
  for (let [key, value] of Object.entries(people)) {
    if (value > 0) {
      totalPeople += value;
      ReactGA.event({
        category: 'Filter',
        action: 'People',
        label: capitalize(key),
        value: value
      });
    }
  }
  ReactGA.event({
    category: 'Filter',
    action: 'People',
    label: 'HouseholdSize',
    value: totalPeople
  });
}

function PeopleFilterModal({ people, setPeople, setShow }) {
  const [localState, setLocalState] = useState(people);
  const { t } = useTranslation();

  return (
    <FilterModal
      className="people"
      show
      onApply={() => {
        setPeople(localState);
        setShow(false);
        sendAnalytics(localState);
      }}
      onHide={() => setShow(false)}
      onClear={() => {
        setShow(false);
        setPeople({});
      }}
      clearDisabled={getTotalPeople(localState) === 0}
      title={t('filterPeople.modalTitle')}
    >
      <PersonInput
        label={t('filterPeople.labelOlderAdults')}
        description={t('filterPeople.descriptionOlderAdults')}
        value={localState.olderAdults || 0}
        onChange={n => setLocalState({ ...localState, olderAdults: n })}
      />

      <PersonInput
        label={t('filterPeople.labelAdults')}
        description={t('filterPeople.descriptionAdults')}
        value={localState.adults || 0}
        onChange={n => setLocalState({ ...localState, adults: n })}
      />

      <PersonInput
        label={t('filterPeople.labelTeens')}
        description={t('filterPeople.descriptionTeens')}
        value={localState.teens || 0}
        onChange={n => setLocalState({ ...localState, teens: n })}
      />

      <PersonInput
        label={t('filterPeople.labelChildren')}
        description={t('filterPeople.descriptionChildren')}
        value={localState.children || 0}
        onChange={n => setLocalState({ ...localState, children: n })}
      />
    </FilterModal>
  );
}

export default function People({ people, setPeople }) {
  const [show, setShow] = useState(false);
  const { t } = useTranslation();
  const totalPeople = getTotalPeople(people);

  return (
    <>
      <Chip
        analytics={{ action: 'People', label: 'Add people' }}
        active={totalPeople > 0}
        onClick={() => setShow(true)}
      >
        {totalPeople > 0
          ? t('filterPeople.buttonWithCount', { count: totalPeople })
          : t('filterPeople.buttonAdd')}
      </Chip>
      {show && <PeopleFilterModal {...{ people, setPeople, setShow }} />}
    </>
  );
}

People.propTypes = {
  people: PropTypes.object.isRequired,
  setPeople: PropTypes.func.isRequired
};
