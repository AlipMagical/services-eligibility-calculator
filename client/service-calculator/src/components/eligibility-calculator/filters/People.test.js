/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import { act } from 'react-dom/test-utils';
import People from './People';
import { axe, toHaveNoViolations } from 'jest-axe';
import { render, fireEvent, waitForElement } from '@testing-library/react';

expect.extend(toHaveNoViolations);

it('renders button without selected people', async () =>
  act(async () => {
    const { container } = render(<People people={{}} setPeople={() => {}} />);

    expect(container).toMatchSnapshot();
    expect(await axe(container)).toHaveNoViolations();
  }));

it('renders modal without selected people', async () =>
  act(async () => {
    const { getByText, getByRole } = render(
      <People people={{}} setPeople={() => {}} />
    );

    fireEvent.click(getByText('Add people'));

    const modal = await waitForElement(() => getByRole('document'));

    expect(modal).toMatchSnapshot();
    expect(await axe(modal)).toHaveNoViolations();
  }));

it('renders button with selected people', async () =>
  act(async () => {
    const { container } = render(
      <People
        people={{ olderAdults: 1, adults: 2, teens: 3, children: 4 }}
        setPeople={() => {}}
      />
    );

    expect(container).toMatchSnapshot();
    expect(await axe(container)).toHaveNoViolations();
  }));

it('renders modal with selected people', async () =>
  act(async () => {
    const { getByText, getByRole } = render(
      <People
        people={{ olderAdults: 1, adults: 2, teens: 3, children: 4 }}
        setPeople={() => {}}
      />
    );

    fireEvent.click(getByText('10 people'));

    const modal = await waitForElement(() => getByRole('document'));

    expect(modal).toMatchSnapshot();
    expect(await axe(modal)).toHaveNoViolations();
  }));
