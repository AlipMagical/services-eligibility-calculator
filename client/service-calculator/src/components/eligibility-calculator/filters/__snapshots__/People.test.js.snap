// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders button with selected people 1`] = `
<div>
  <button
    class="btn btn-chip ellipsis active"
  >
    10 people
  </button>
</div>
`;

exports[`renders button without selected people 1`] = `
<div>
  <button
    class="btn btn-chip ellipsis "
  >
    Add people
  </button>
</div>
`;

exports[`renders modal with selected people 1`] = `
<div
  class="modal-dialog modal-md modal-dialog-centered"
  role="document"
>
  <div
    class="modal-content"
  >
    <div
      class="modal-header"
    >
      <div
        class="text-center w-100 text-center modal-title h4"
      >
        People in your household
        <button
          aria-label="Close"
          class="close float-right modal-close"
          data-dismiss="modal"
          type="button"
        >
          <span
            aria-hidden="true"
            class="align-middle"
          >
            ×
          </span>
        </button>
      </div>
    </div>
    <div
      class="modal-body"
    >
      <div
        class="my-3 row"
      >
        <div
          class="col-6"
        >
          <span
            class="input-label"
          >
            Older adults
          </span>
          <span
            class="people-input-description"
          >
            Ages 59+
          </span>
        </div>
        <div
          class="people-input col-6"
        >
          <button
            aria-label="Decrease Older adults"
            class="btn btn-people-input"
          >
            <i
              class="material-icons align-middle"
            >
              remove
            </i>
          </button>
          <span
            class="people-input-value align-middle"
          >
            1
          </span>
          <button
            aria-label="Increase Older adults"
            class="btn btn-people-input"
          >
            <i
              class="material-icons align-middle"
            >
              add
            </i>
          </button>
        </div>
      </div>
      <div
        class="my-3 row"
      >
        <div
          class="col-6"
        >
          <span
            class="input-label"
          >
            Adults
          </span>
          <span
            class="people-input-description"
          >
            Ages 18-58
          </span>
        </div>
        <div
          class="people-input col-6"
        >
          <button
            aria-label="Decrease Adults"
            class="btn btn-people-input"
          >
            <i
              class="material-icons align-middle"
            >
              remove
            </i>
          </button>
          <span
            class="people-input-value align-middle"
          >
            2
          </span>
          <button
            aria-label="Increase Adults"
            class="btn btn-people-input"
          >
            <i
              class="material-icons align-middle"
            >
              add
            </i>
          </button>
        </div>
      </div>
      <div
        class="my-3 row"
      >
        <div
          class="col-6"
        >
          <span
            class="input-label"
          >
            Teens
          </span>
          <span
            class="people-input-description"
          >
            Ages 13-17
          </span>
        </div>
        <div
          class="people-input col-6"
        >
          <button
            aria-label="Decrease Teens"
            class="btn btn-people-input"
          >
            <i
              class="material-icons align-middle"
            >
              remove
            </i>
          </button>
          <span
            class="people-input-value align-middle"
          >
            3
          </span>
          <button
            aria-label="Increase Teens"
            class="btn btn-people-input"
          >
            <i
              class="material-icons align-middle"
            >
              add
            </i>
          </button>
        </div>
      </div>
      <div
        class="my-3 row"
      >
        <div
          class="col-6"
        >
          <span
            class="input-label"
          >
            Children
          </span>
          <span
            class="people-input-description"
          >
            Ages 0-12
          </span>
        </div>
        <div
          class="people-input col-6"
        >
          <button
            aria-label="Decrease Children"
            class="btn btn-people-input"
          >
            <i
              class="material-icons align-middle"
            >
              remove
            </i>
          </button>
          <span
            class="people-input-value align-middle"
          >
            4
          </span>
          <button
            aria-label="Increase Children"
            class="btn btn-people-input"
          >
            <i
              class="material-icons align-middle"
            >
              add
            </i>
          </button>
        </div>
      </div>
    </div>
    <div
      class="modal-footer"
    >
      <button
        class="btn btn-clear btn-gutter my-0"
      >
        Clear
      </button>
      <button
        class="btn btn-primary float-right my-0 mx-0"
      >
        Apply
      </button>
    </div>
  </div>
</div>
`;

exports[`renders modal without selected people 1`] = `
<div
  class="modal-dialog modal-md modal-dialog-centered"
  role="document"
>
  <div
    class="modal-content"
  >
    <div
      class="modal-header"
    >
      <div
        class="text-center w-100 text-center modal-title h4"
      >
        People in your household
        <button
          aria-label="Close"
          class="close float-right modal-close"
          data-dismiss="modal"
          type="button"
        >
          <span
            aria-hidden="true"
            class="align-middle"
          >
            ×
          </span>
        </button>
      </div>
    </div>
    <div
      class="modal-body"
    >
      <div
        class="my-3 row"
      >
        <div
          class="col-6"
        >
          <span
            class="input-label"
          >
            Older adults
          </span>
          <span
            class="people-input-description"
          >
            Ages 59+
          </span>
        </div>
        <div
          class="people-input col-6"
        >
          <button
            aria-label="Decrease Older adults"
            class="btn btn-people-input"
            disabled=""
          >
            <i
              class="material-icons align-middle"
            >
              remove
            </i>
          </button>
          <span
            class="people-input-value align-middle"
          >
            0
          </span>
          <button
            aria-label="Increase Older adults"
            class="btn btn-people-input"
          >
            <i
              class="material-icons align-middle"
            >
              add
            </i>
          </button>
        </div>
      </div>
      <div
        class="my-3 row"
      >
        <div
          class="col-6"
        >
          <span
            class="input-label"
          >
            Adults
          </span>
          <span
            class="people-input-description"
          >
            Ages 18-58
          </span>
        </div>
        <div
          class="people-input col-6"
        >
          <button
            aria-label="Decrease Adults"
            class="btn btn-people-input"
            disabled=""
          >
            <i
              class="material-icons align-middle"
            >
              remove
            </i>
          </button>
          <span
            class="people-input-value align-middle"
          >
            0
          </span>
          <button
            aria-label="Increase Adults"
            class="btn btn-people-input"
          >
            <i
              class="material-icons align-middle"
            >
              add
            </i>
          </button>
        </div>
      </div>
      <div
        class="my-3 row"
      >
        <div
          class="col-6"
        >
          <span
            class="input-label"
          >
            Teens
          </span>
          <span
            class="people-input-description"
          >
            Ages 13-17
          </span>
        </div>
        <div
          class="people-input col-6"
        >
          <button
            aria-label="Decrease Teens"
            class="btn btn-people-input"
            disabled=""
          >
            <i
              class="material-icons align-middle"
            >
              remove
            </i>
          </button>
          <span
            class="people-input-value align-middle"
          >
            0
          </span>
          <button
            aria-label="Increase Teens"
            class="btn btn-people-input"
          >
            <i
              class="material-icons align-middle"
            >
              add
            </i>
          </button>
        </div>
      </div>
      <div
        class="my-3 row"
      >
        <div
          class="col-6"
        >
          <span
            class="input-label"
          >
            Children
          </span>
          <span
            class="people-input-description"
          >
            Ages 0-12
          </span>
        </div>
        <div
          class="people-input col-6"
        >
          <button
            aria-label="Decrease Children"
            class="btn btn-people-input"
            disabled=""
          >
            <i
              class="material-icons align-middle"
            >
              remove
            </i>
          </button>
          <span
            class="people-input-value align-middle"
          >
            0
          </span>
          <button
            aria-label="Increase Children"
            class="btn btn-people-input"
          >
            <i
              class="material-icons align-middle"
            >
              add
            </i>
          </button>
        </div>
      </div>
    </div>
    <div
      class="modal-footer"
    >
      <button
        class="disabled btn btn-clear btn-gutter my-0"
      >
        Clear
      </button>
      <button
        class="btn btn-primary float-right my-0 mx-0"
      >
        Apply
      </button>
    </div>
  </div>
</div>
`;
